File: lintcode/easy/delete_node_in_the_middle_of_singly_linked_list/py/delete_node_in_the_middle_of_singly_linked_list.py
# coding:utf-8
'''
@Copyright:LintCode
@Author:   lilsweetcaligula
@Problem:  http://www.lintcode.com/problem/delete-node-in-the-middle-of-singly-linked-list
@Language: Python
@Datetime: 17-02-15 20:40
'''

"""
Definition of ListNode
class ListNode(object):

    def __init__(self, val, next=None):
        self.val = val
        self.next = next
"""
class Solution:
    # @param node: the node in the list should be deleted
    # @return: nothing
    def deleteNode(self, node):
        if node == None or node.next == None:
            return
        
        node.val  = node.next.val
        node.next = node.next.next

File: lintcode/easy/delete_node_in_the_middle_of_singly_linked_list/cpp/delete_node_in_the_middle_of_singly_linked_list.cpp
/*
@Copyright:LintCode
@Author:   lilsweetcaligula
@Problem:  http://www.lintcode.com/problem/delete-node-in-the-middle-of-singly-linked-list
@Language: C++
@Datetime: 17-02-15 20:37
*/

/**
 * Definition of ListNode
 * class ListNode {
 * public:
 *     int val;
 *     ListNode *next;
 *     ListNode(int val) {
 *         this->val = val;
 *         this->next = NULL;
 *     }
 * }
 */
class Solution {
public:
    /**
     * @param node: a node in the list should be deleted
     * @return: nothing
     */
    void deleteNode(ListNode *node) 
    {
        auto temp = node->next;
        *node     = std::move(*node->next);
        
        delete temp; temp = nullptr;
    }
};